dhall-json/src/Dhall/JSONToDhall.hs

```diff
@@ -369,13 +380,257 @@ typeCheckSchemaExpr compileException expr =
       D.Const D.Type -> return expr
       _              -> throwM . compileException $ BadDhallType t expr
 
-keyValMay :: A.Value -> Maybe (Text, A.Value)
+keyValMay :: Value -> Maybe (Text, Value)
 keyValMay (A.Object o) = do
      A.String k <- HM.lookup "key" o
      v <- HM.lookup "value" o
      return (k, v)
 keyValMay _ = Nothing
 
+{-| Given a JSON `Value`, make a best-effort guess of what the matching Dhall
+    type should be
+
+    This is used by @{json,yaml}-to-dhall@ if the user does not supply a schema
+    on the command line
+-}
+inferSchema :: Value -> Schema
+inferSchema (A.Object m) =
+    let convertMap = Data.Map.fromList . HM.toList
+
+    in (Record . RecordSchema . convertMap) (fmap inferSchema m)
+inferSchema (A.Array xs) =
+    List (Foldable.fold (fmap inferSchema xs))
+inferSchema (A.String _) =
+    Text
+inferSchema (A.Number n) =
+    case floatingOrInteger n of
+        Left (_ :: Double) -> Double
+        Right (integer :: Integer)
+            | 0 <= integer -> Natural
+            | otherwise    -> Integer
+inferSchema (A.Bool _) = do
+    Bool
+inferSchema A.Null = do
```

Comment:
What do these `do`s do? I don't even understand how this typechecks!

Reply:
@sjakobi: They are an accidental remnant of an earlier draft of this code that had an impure `inferSchema` function before I realized it could be made pure.  They are not necessary and I will remove them.

To answer you question: you can put a `do` around literally any expression in Haskell because `do x` desugars to `x`.  For example, this is valid Haskell code:

```haskell
(do 1) + (do 1)
```

This works because a `do` block with a single line doesn't desugar to the use of the `>>=` operator, so there is no `Monad` constraint for such an expression!  That means that this is also legal:

```haskell
example :: Integer
example = do
    let x = 1

    let y = 1

    x + y
```

Some people even intentionally use `do` in this way (although I'm not one of them).

Reply:
Great explanation! Thanks!

Reply:
You're welcome! 🙂